storage capacity tests

pohly opened the pull request:
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:

This adds tests for https://github.com/kubernetes/kubernetes/issues/72031.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

pohly committed:
mock tests: ResourceExhausted error handling in external-provisioner

The mock driver gets instructed to return a ResourceExhausted error
for the first CreateVolume invocation via the storage class
parameters.

How this should be handled depends on the situation: for normal
volumes, we just want external-scheduler to retry. For late binding,
we want to reschedule the pod. It also depends on topology support.

pohly commented:
WIP because csi-test and external-provisioner need to be updated for all test cases to pass.

pohly committed:
mock tests: less verbose log output checking

parseMockLogs is called potentially multiple times while waiting for
output. Dumping all CSI calls each time is quite verbose and
repetitive. To verify what the driver has done already, the normal
capturing of the container log can be used instead:

csi-mockplugin-0/mock@127.0.0.1: gRPCCall: {"Method":"/csi.v1.Node/NodePublishVolume","Request"...

pohly committed:
mock tests: remove redundant retrieval of log output

The code became obsolete with the introduction of parseMockLogs
because that will retrieve the log itself. For debugging of a running
test the normal pod output logging is sufficient.

pohly committed:
mock tests: add timeout

The for loop that waited for the signal to delete pod had no timeout,
so if something went wrong, it would wait for the entire test suite to
time out.

pohly committed:
mock tests: remove redundant wrapping of error

The "error waiting for expected CSI calls" is redundant because it's
immediately followed by checking that error with:

   framework.ExpectNoError(err, "while waiting for all CSI calls")

pohly commented:
I have rebased on top of https://github.com/kubernetes/kubernetes/pull/89041 to reuse some of the new utility functions in that PR (specifically, `compareCSICalls`) and to avoid conflicts when that PR merges.

In addition, I am trying some enhancements for that PR.

pohly commented:
/retest

Failures were unrelated.

pohly commented:
/retest

pohly commented:
@jsafrane what do you think about the log output retry approach (d6f0d0e)? Do you want to copy that into your PR or shall we merge your changes together with mine in this PR?

jsafrane commented:
@pohly, thanks for the suggestion. I incorporated your last commit into #72031 (with you as author) to see if it helps.

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
@jsafrane: I've rebased on top of your merged PR and the tests passed now after a few flakes. Perhaps you can have a look, as you are familiar with code and I am touching a few things you just worked on?

jsafrane commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -730,6 +743,202 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
+				}
+
+				var calls []mockCSICall
+				err = wait.Poll(time.Second, csiPodRunningTimeout, func() (done bool, err error) {
```
By this time, test pvc/pod has been created, run and deleted. So all calls must be in mock driver logs. Isn't 5 minute timeout too much to get the logs?

pohly replied:
I've replaced this with a per-test timeout. It would be even nicer if that deadline could also be passed into the other helper functions (like `WaitForPodNameRunningInNamespace`) but those have their own builtin timeouts.

@jsafrane: please have another look

pohly committed:
mock tests: per-test timeout for ResourceExhausted

The timeout for the two loops inside the test itself are now bounded
by an upper limit for the duration of the entire test instead of
having their own, rather arbitrary timeouts.

pohly commented:
/retest

jsafrane commented:
/lgtm
/approve

pohly commented:
/retest
